imports/lib/models/Model.ts

```diff
@@ -280,6 +282,70 @@ export function formatValidationError(error: unknown) {
   error.stack = error.stack?.replace(/^.*$/m, error.message);
 }
 
+export type NormalizedIndexSpecification = readonly [string, IndexDirection][];
+
+const AllowedIndexOptions = ['unique', 'sparse', 'partialFilterExpression', 'expireAfterSeconds'] as const;
+type AllowedIndexOptionsType = typeof AllowedIndexOptions[number];
+export type IndexOptions = Pick<CreateIndexesOptions, AllowedIndexOptionsType>;
+type NormalizedIndexOptions = [AllowedIndexOptionsType, any][];
+
+export interface ModelIndexSpecification {
+  index: NormalizedIndexSpecification;
+  options: NormalizedIndexOptions;
+  // JSON-serialize [index, options] to make it easier to compare against
+  // existing indexes
+  stringified: string;
+}
+
+function indexIsField(index: any): index is string {
+  return typeof index === 'string';
+}
+
+function indexIsFieldDirectionPair(index: any): index is readonly [string, IndexDirection] {
+  return Array.isArray(index) && index.length === 2 && typeof index[0] === 'string';
+}
+
+function indexIsObject(index: any): index is Record<string, IndexDirection> {
+  return typeof index === 'object' && index !== null && !Array.isArray(index);
+}
+
+function indexIsMap(index: any): index is Map<string, IndexDirection> {
+  return index instanceof Map;
+}
+
+export function normalizeIndexSpecification(
+  index: IndexSpecification
+): NormalizedIndexSpecification {
+  // An index specification can be a:
+  // - string
+  // - array with the first element as field name and the second element as
+  //   direction
+  // - object with keys as field names and values as direction (note that ES2015
+  //   and later preserves object insertion order for string keys)
+  // - Map with keys as field names and values as direction
+  // - An array mix-and-matching any of the above So we need to detect each of
+  //   them and normalize to a single format (array of arrays)
+  if (indexIsField(index)) {
+    return [[index, 1]];
+  } else if (indexIsFieldDirectionPair(index)) {
+    return [index];
+  } else if (indexIsObject(index)) {
+    return Object.entries(index);
+  } else if (indexIsMap(index)) {
+    return [...index];
+  } else {
+    return index.map(normalizeIndexSpecification).reduce((acc, val) => acc.concat(val), []);
+  }
+}
+
+export function normalizeIndexOptions(options: CreateIndexesOptions): NormalizedIndexOptions {
```

Comment:
Rather than filter out unsupported options here, I'd prefer you threw an exception to make it loud and clear that the option is unsupported.

Reply:
Ahhh, but now that I'm reading the part that retrieves current indexes from the DB, I see that you have to handle whatever is in the database.  Maybe have a flag or a separate function for the two behaviors, so as to explicitly disallow unsupported `addIndex` calls, but be tolerant of whatever's in the DB?

Reply:
I guess the type declaration prevents incorrect usage of `addIndex` already.  Never mind!